File: ios/PSMobile/UI/PSScale.swift
import SwiftUI
import PSMShared

/// Wie stark die Oberflaeche insgesamt verkleinert wird.
///
/// Die Zahlen kommen aus dem gemeinsamen Modul (E-13) - es sind
/// buchstaeblich dieselben wie auf Android, nicht nachgehaltene Kopien. Die Masse in den Bildschirmen sind
/// fuer ein Tablet mit rund 1000 x 720 Punkt geschrieben. Faellt das
/// Fenster kleiner aus - iPhone, Slide Over, Split View, Stage Manager -,
/// dann sind dieselben Masse im Verhaeltnis zu gross.
///
/// Auf Android wird dafuer die wirksame Dichte gestaucht. SwiftUI hat
/// kein Gegenstueck dazu, das global wirkt, deshalb rechnen die
/// Bildschirme ihre Masse hier durch: `ps.pt(16)` statt `16`. Das ist
/// mehr Schreibarbeit, aber es ist dieselbe Regel - und sie ist an einer
/// Stelle nachzulesen statt in fuenfzehn Ansichten verstreut.
///
/// Wer eine Zahl direkt hinschreibt, bekommt sie auf dem iPhone zu gross
/// zurueck. Genau davor schuetzt diese Datei.
struct PSScale: Equatable {

    /// Die Groesse, fuer die die Masse gedacht sind.
    static let referenceWidth = CGFloat(WindowScale.shared.REFERENCE_WIDTH)
    static let referenceHeight = CGFloat(WindowScale.shared.REFERENCE_HEIGHT)

    /// Untergrenze. Darunter waeren Zielflaechen physisch zu klein zum
    /// Treffen. Wo es enger wird, muss der Bildschirm selbst Inhalt
    /// weglassen, statt weiter zu schrumpfen.
    static let minScale = CGFloat(WindowScale.shared.MIN_SCALE)

    /// Wie stark die Schrift dem Kastenmass folgt. Text darf nicht so
    /// stark schrumpfen wie Kaesten, sonst wird er unleserlich, bevor der
    /// Platz wirklich knapp ist.
    static let fontFollow = CGFloat(WindowScale.shared.FONT_FOLLOW)

    let factor: CGFloat

    init(width: CGFloat, height: CGFloat) {
        self.factor = PSScale.scaleFor(width: width, height: height)
    }

    /// Die knappere Kante entscheidet. Hochskaliert wird nie: ab der
    /// Referenzgroesse stimmen die Masse bereits.
    static func scaleFor(width: CGFloat, height: CGFloat) -> CGFloat {
        CGFloat(WindowScale.shared.forWindow(width: Float(width), height: Float(height)))
    }

    /// Gedaempfte Fassung fuer Schriftgroessen.
    static func fontScaleFor(_ scale: CGFloat) -> CGFloat {
        CGFloat(WindowScale.shared.fontScale(scale: Float(scale)))
    }

    /// Ein Kastenmass: Abstand, Breite, Hoehe, Eckenradius.
    func pt(_ value: CGFloat) -> CGFloat { value * factor }

    /// Eine Schriftgroesse - schrumpft schwaecher als die Kaesten.
    func font(_ value: CGFloat) -> CGFloat {
        value * PSScale.fontScaleFor(factor)
    }

    /// Zielflaeche fuer Finger und Stift. Bewusst nicht kleiner als
    /// Apples Untergrenze von 44 Punkt, egal was die Skalierung sagt.
    func touch(_ value: CGFloat = 44) -> CGFloat { max(pt(value), 44) }
}

private struct PSScaleKey: EnvironmentKey {
    /// Falls jemand die Umgebung vergisst: die Referenzgroesse, also
    /// unveraendert. Lieber zu gross auf einem grossen Geraet als
    /// willkuerlich verkleinert.
    static let defaultValue = PSScale(width: PSScale.referenceWidth,
                                      height: PSScale.referenceHeight)
}

extension EnvironmentValues {
    var psScale: PSScale {
        get { self[PSScaleKey.self] }
        set { self[PSScaleKey.self] = newValue }
    }
}

/// Legt die Skalierung aus der tatsaechlichen Fenstergroesse fest.
///
/// Bewusst die Fenstergroesse und nicht die Bildschirmgroesse: Auf dem
/// iPad kann die App in Slide Over oder Stage Manager in einem schmalen
/// Fenster stehen, waehrend der Bildschirm gross bleibt. Wer den
/// Bildschirm misst, baut dort eine Oberflaeche, die nicht hineinpasst.
struct PSScaleRoot<Content: View>: View {
    @ViewBuilder let content: () -> Content

    var body: some View {
        GeometryReader { geo in
            content()
                .environment(\.psScale,
                             PSScale(width: geo.size.width, height: geo.size.height))
        }
    }
}

File: android/app/src/main/java/de/psmobile/ui/PsScale.kt
package de.psmobile.ui

import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.DpSize
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import de.psmobile.shared.ui.WindowScale
import de.psmobile.ui.theme.psTouch

/**
 * Gegenstueck zu `PSScale` in `ios/PSMobile/UI/PSScale.swift`.
 *
 * iOS rechnet jedes Mass einzeln durch (`ps.pt(16)`), Android staucht
 * die wirksame Dichte global in [de.psmobile.ui.theme.PSMobileTheme].
 * Beides ist dieselbe Regel aus dem gemeinsamen Modul (`WindowScale`).
 * Damit die portierten Bildschirme Zeile fuer Zeile wie ihr Swift-
 * Original aussehen, gibt es die Aufrufe hier trotzdem - `pt` liefert
 * schlicht den dp-Wert, den das Theme dann staucht.
 */
@Immutable
data class PsScale(
    /** Wie stark das Theme staucht - 1 ab der Referenzgroesse. */
    val factor: Float,
    /**
     * Die Fenstergroesse im selben Massystem wie [pt]: Wer auf iOS
     * `min(ps.pt(380), ps.windowSize.width * 0.4)` schreibt, schreibt
     * hier dasselbe.
     */
    val windowSize: DpSize,
) {
    /** Ein Kastenmass: Abstand, Breite, Hoehe, Eckenradius. */
    fun pt(value: Int): Dp = value.dp
    fun pt(value: Float): Dp = value.dp
    fun pt(value: Double): Dp = value.toFloat().dp

    /** Eine Schriftgroesse - das Theme laesst sie schwaecher schrumpfen. */
    fun font(value: Int): TextUnit = value.sp
    fun font(value: Float): TextUnit = value.sp
    fun font(value: Double): TextUnit = value.toFloat().sp

    /**
     * Zielflaeche fuer Finger und Stift - nie kleiner als 44 physische
     * dp, egal was die Stauchung sagt. Siehe [psTouch].
     */
    fun touch(value: Int = 44): Dp {
        if (factor >= 1f) return value.dp
        val gewuenscht = maxOf(value * factor, 44f)
        return (gewuenscht / factor).dp
    }
    fun touch(value: Float): Dp = touch(value.toInt())

    companion object {
        val referenceWidth: Float = WindowScale.REFERENCE_WIDTH
        val referenceHeight: Float = WindowScale.REFERENCE_HEIGHT
        val minScale: Float = WindowScale.MIN_SCALE
        val fontFollow: Float = WindowScale.FONT_FOLLOW

        fun scaleFor(width: Float, height: Float): Float = WindowScale.forWindow(width, height)
        fun fontScaleFor(scale: Float): Float = WindowScale.fontScale(scale)

        /** Ohne Umgebung: die Referenzgroesse, also unveraendert. */
        val Default = PsScale(1f, DpSize(referenceWidth.dp, referenceHeight.dp))
    }
}

val LocalPsScale = compositionLocalOf { PsScale.Default }

/**
 * Legt die Skalierung aus der tatsaechlichen Fenstergroesse fest -
 * Gegenstueck zu `PSScaleRoot`. Muss innerhalb von [PSMobileTheme]
 * stehen, denn die Fenstergroesse wird hier im gestauchten Massystem
 * gemessen (so wie die Bildschirme rechnen).
 */
@Composable
fun PSScaleRoot(content: @Composable () -> Unit) {
    val configuration = LocalConfiguration.current
    val factor = remember(configuration.screenWidthDp, configuration.screenHeightDp) {
        PsScale.scaleFor(configuration.screenWidthDp.toFloat(), configuration.screenHeightDp.toFloat())
    }
    BoxWithConstraints(Modifier) {
        val scale = remember(factor, maxWidth, maxHeight) {
            PsScale(factor, DpSize(maxWidth, maxHeight))
        }
        CompositionLocalProvider(LocalPsScale provides scale, content = content)
    }
}
